SME feedback

MaiHameed opened the pull request:
Addressed #10
- Rewrote the first paragraph to jump into the concept of acknowledgement quickly.
- Rewrote the `MicroProfile Reactive Messaging Acknowledgement Strategies` section completely and removed the table. Started with an explanation of acknowledgement, leads into the default strategies depending on method signature, explains `POST_PROCESSING`, explains why it's not sufficient for the `system` service to lead into the next section that implements the `MANUAL` strategy instead. `NONE` and `PRE_PROCESSING` were not explained, and we decided not to try and explain general best practice, and instead just go into the strategies used in this specific use case and why they're used.
- Added the suggested paragraphs and comments taken from #10 into README and code with slight modifications.
- Rewrote the `Testing the application` section to match the rest integration guide with a focus on the newly implemented acknowledgement features.

Most recent version of the guide can be found here:  http://lgdev1.fyre.ibm.com:4020/guides/microprofile-reactive-messaging-acknowledgement.html

Azquelt commented on README.adoc:
```diff
@@ -31,9 +31,8 @@ Learn how to acknowledge messages using MicroProfile Reactive Messaging.
-Acknowledgement differentiates between a message that
+Acknowledgements are a key concept in MicroProfile Reactive Messaging which ensure that messages are not lost in the
+event of a system failure. Acknowledgement differentiates between a message that
 has been delivered to the target service, and a message that has been processed and completed. Every message that gets sent
```
I don't think it's clear what "completed" means in this case.

Essentially we mean "we don't want the message delivered again in the case of system failure" but that's a bit wordy.

Maybe we could say something like this: "Acknowledgement differentiates between a message has been delivered to the target service, and a message which has been fully processed."

Later, we can go into more detail about what "fully processed" means.

MaiHameed replied:
I updated this paragraph when addressing some of Laura's feedback. See lines 34-37 here: https://github.com/OpenLiberty/draft-guide-microprofile-reactive-messaging-acknowledgement/commit/40edf9c64e9f416f39a021caf501037838006e08#diff-7bd4a925c695c2eb0eced3872b9b965fR34-R37.

Azquelt commented on README.adoc:
```diff
@@ -61,79 +60,74 @@ include::{common-includes}/gitclone.adoc[]
+implicitly by the MicroProfile Reactive Messaging implementation. Acknowledgement for the incoming messages is controlled
+by the `@Acknowledgment` annotation in the MicroProfile Reactive Messaging specification. If the `@Acknowledgement`
+annotation is not defined explicitly, the default acknowledgement strategy applies, and depends on the method signature.
+Only methods that receive incoming messages and are annotated with `@Incoming` need to acknowledge messages, methods
+that are only annotated with `@Outgoing` methods do not support acknowledgement.
```
> methods that are only annotated with `@Outgoing` methods do not support acknowledgement

I think it's helpful to point out that only messages annotated with `@Incoming` need to worry about acknowledgement, but I think it's misleading to say that "`@Outgoing` methods do not support acknowledgement".

It's not that they don't support it, it's that the requirement is that you ensure that every message you _receive_ is acknowledged, and so this requirement only applies to methods annotated with `@Incoming`.

Azquelt replied:
The spec does actually use similar wording which isn't great here: https://download.eclipse.org/microprofile/microprofile-reactive-messaging-1.0/microprofile-reactive-messaging-spec.html#_message_acknowledgement

However it does explain that the reason they don't support acknowledgement is because they don't receive an input message.

MaiHameed replied:
Addressing comments from here into new PR #13, this specific comment is addressed here https://github.com/OpenLiberty/draft-guide-microprofile-reactive-messaging-acknowledgement/pull/13/commits/6e00592a46e342f4583e8e10188c56258e87cebb

Azquelt commented on README.adoc:
```diff
@@ -61,79 +60,74 @@ include::{common-includes}/gitclone.adoc[]
+that are only annotated with `@Outgoing` methods do not support acknowledgement.
+
+Different strategy defaults and supported acknowledgement for each supported signature could be found at the
+https://openliberty.io/docs/latest/reference/javadoc/microprofile-3.3-javadoc.html#package=org/eclipse/microprofile/reactive/messaging/package-frame.html&class=overview-summary.html[MicroProfile Reactive Messaging Javadoc^].
```
I couldn't find the list of the supported acknowledgement strategies and the defaults for each signature in the javadoc. I know they are listed in the spec here: https://download.eclipse.org/microprofile/microprofile-reactive-messaging-1.0/microprofile-reactive-messaging-spec.html#_message_acknowledgement

MaiHameed replied:
Yep, unfortunately the Javadocs on the OL website aren't very helpful. I raised an issue to address it here: https://github.com/OpenLiberty/docs/issues/2371. I'm also addressing Laura's comments from issue #12 and we decided to just strip the sentence out completely since it's not necessary for the user to know in order to complete the guide.

Azquelt commented on README.adoc:
```diff
@@ -61,79 +60,74 @@ include::{common-includes}/gitclone.adoc[]
-                        by any other existing Reactive Messaging implementation, and is often used when the incoming message
-                        should be acknowledged after the outgoing message is acknowledged.
-|===
+By default, all of the methods that require acknowledging messages are assigned the `POST_PROCESSING` strategy. If the
+acknowledgement strategy is set to `POST_PROCESSING`, the Reactive Messaging implementation will acknowledge the message
```
This section is not quite correct.

1. The default acknowledgement strategy depends on the method signature
1. Additionally, the spec isn't even consistent in the way it specifies the default (this really sucks, but we're fixing that in the next spec release)

MaiHameed replied:
My mistake, good catch, fixed with https://github.com/OpenLiberty/draft-guide-microprofile-reactive-messaging-acknowledgement/pull/13/commits/38441cd33d674ddcfde709bb9211e94c281047ab

Azquelt commented on README.adoc:
```diff
@@ -61,79 +60,74 @@ include::{common-includes}/gitclone.adoc[]
-                        by any other existing Reactive Messaging implementation, and is often used when the incoming message
-                        should be acknowledged after the outgoing message is acknowledged.
-|===
+By default, all of the methods that require acknowledging messages are assigned the `POST_PROCESSING` strategy. If the
+acknowledgement strategy is set to `POST_PROCESSING`, the Reactive Messaging implementation will acknowledge the message
+depending on if the annotated method emits data:
+
+    1. If the method emits data, the incoming message will be acknowledged once the outgoing message is acknowledged.
+    2. If the method does not emit data, the incoming message will be acknowledged once the method or processing completes.
```
I'm not sure whether it would make sense here to explain the qualities we're looking for why how that matches up with the `POST_PROCESSING` strategy.

MaiHameed replied:
Sorry, I don't understand what you mean here?

Azquelt replied:
### Reliable messaging

In most use cases, our messages are important and we need to ensure that they are all processed.

To ensure that each message is processed fully at least once, we have the following requirement:
* don't acknowledge the message until all processing of that message is complete

(we could refine that requirement a bit more to explain what we mean by complete)

Using the post processing acknowledgement strategy at every processing step will fulfil this requirement, which is why we want to use it.

Where we can't use the post processing acknowledgement strategy, we should use the manual strategy _with the goal of fulfilling the same requirement_.

### Unreliable messaging

In other circumstances, we may not care so much if we lose a few messages.

In this case, we do not need to ensure that each message is processed fully at least once, so we don't care about this requirement, so using the pre-processing strategy might be appropriate.

MaiHameed replied:
Added a small paragraph here: https://github.com/OpenLiberty/draft-guide-microprofile-reactive-messaging-acknowledgement/pull/13/commits/f3d9a6de56e07967412dae22ef36e66d6d23ac40

Azquelt commented on README.adoc:
```diff
@@ -61,79 +60,74 @@ include::{common-includes}/gitclone.adoc[]
+    1. If the method emits data, the incoming message will be acknowledged once the outgoing message is acknowledged.
+    2. If the method does not emit data, the incoming message will be acknowledged once the method or processing completes.
+
+The acknowledgement strategy to use for a method depends on the use case of said method. The `system` service contains
+a method which defaults to the `POST_PROCESSING` strategy. This is insufficient and does not produce the desired behaviour.
```
This part is wrong. If the method signature used supports the `POST_PROCESSING` strategy, then there's no reason not to use it.

However, the method used in the `system` service is
```java
    @Incoming("propertyRequest")
    @Outgoing("propertyResponse")
    public PropertyMessage sendProperty(String propertyName) {
        logger.info("sendProperty: " + propertyName);
        String propertyValue = System.getProperty(propertyName);
        if (propertyValue == null) {
            logger.warning(propertyName + " is not System property.");
            return null;
        }
        return new PropertyMessage(getHostname(),
                propertyName,
                System.getProperty(propertyName, "unknown"));
    }
```

This method is incorrectly implemented because it can return `null` which is not valid.

It should be something like this:
```java
    @Incoming("propertyRequest")
    @Outgoing("propertyResponse")
    public PublisherBuilder<PropertyMessage> sendProperty(String propertyName) {
        logger.info("sendProperty: " + propertyName);
        String propertyValue = System.getProperty(propertyName);
        if (propertyValue == null) {
            logger.warning(propertyName + " is not System property.");
            return ReactiveStreams.empty();
        }
        PropertyMessage message = new PropertyMessage(getHostname(),
                propertyName,
                System.getProperty(propertyName, "unknown"));
        return ReactiveStreams.of(message);
    }
```

Checking the spec, this new signature matches up with
```java
@Incoming("in")
@Outgoing("out")
PublisherBuilder<O> method(I payload)
```

This signature does not support post-processing acknowledgement, which is why manual acknowledgement is required.

MaiHameed replied:
Fixed here: https://github.com/OpenLiberty/draft-guide-microprofile-reactive-messaging-acknowledgement/pull/13/commits/709306b48b5fc5b21a7ddd91d5800818c280fa4d

Azquelt commented on README.adoc:
```diff
@@ -61,79 +60,74 @@ include::{common-includes}/gitclone.adoc[]
+property [hotspot=null file=0]`doesn't exist`, the request is ignored and no property response is
+[hotspot=returnNull file=0]`returned`.
+
+There is a key difference here, in the case where a property response is produced, the request is not finished processing
+until the response has been sent and safely stored, and it is only then that the incoming message should be acknowledged.
```
```suggestion
There is a key difference here, in the case where a property response is produced, the request is not finished processing
until the response has been sent and safely stored by the Kafka broker, and it is only then that the incoming message should be acknowledged.
```

MaiHameed replied:
Fixed https://github.com/OpenLiberty/draft-guide-microprofile-reactive-messaging-acknowledgement/pull/13/commits/7ff1468ec3eab14c9fd53bfb8a64d33c24173cb7

MaiHameed commented:
I addressed all the review comments @Azquelt :) Updated the live site [here](http://lgdev1.fyre.ibm.com:4020/guides/microprofile-reactive-messaging-acknowledgement.html) if you'd like to take a look at it

MaiHameed commented:
All latest changes will be addressed with #13
